packages/mui-base/src/unstable_useModal/useModal.ts

```diff
@@ -0,0 +1,239 @@
+  UseModalReturnValue,
+  UseModalRootSlotProps,
+  UseModalBackdropSlotProps,
+} from './useModal.types';
+
+function getContainer(container: UseModalParameters['container']) {
+  return typeof container === 'function' ? container() : container;
+}
+
+function getHasTransition(children: UseModalParameters['children']) {
+  return children ? children.props.hasOwnProperty('in') : false;
+}
+
+// A modal manager used to track and manage the state of open Modals.
+// Modals don't open on the server so this won't conflict with concurrent requests.
+const defaultManager = new ModalManager();
+/**
+ *
+ * Demos:
+ *
+ * - [Modal](https://mui.com/base-ui/react-modal/#hook)
+ *
+ * API:
+ *
+ * - [useModal API](https://mui.com/base-ui/react-modal/hooks-api/#use-modal)
+ */
+export function useModal(parameters: UseModalParameters): UseModalReturnValue {
+  const {
+    container,
+    disableEscapeKeyDown = false,
+    disableScrollLock = false,
+    // @ts-ignore internal logic - Base UI supports the manager as a prop too
+    manager = defaultManager,
+    closeAfterTransition = false,
+    onTransitionEnter,
+    onTransitionExited,
+    children,
+    onClose,
+    open,
+    rootRef,
+  } = parameters;
+
+  // @ts-ignore internal logic
+  const modal = React.useRef<{ modalRef: HTMLDivElement; mount: HTMLElement }>({});
+  const mountNodeRef = React.useRef<null | HTMLElement>(null);
+  const modalRef = React.useRef<null | HTMLDivElement>(null);
+  const handleRef = useForkRef(modalRef, rootRef);
+  const [exited, setExited] = React.useState(!open);
+  const hasTransition = getHasTransition(children);
+
+  let ariaHiddenProp = true;
+  if (
+    parameters['aria-hidden'] === 'false' ||
+    (typeof parameters['aria-hidden'] === 'boolean' && !parameters['aria-hidden'])
+  ) {
+    ariaHiddenProp = false;
+  }
+
+  const getDoc = () => ownerDocument(mountNodeRef.current);
+  const getModal = () => {
+    modal.current.modalRef = modalRef.current as HTMLDivElement;
+    modal.current.mount = mountNodeRef.current as HTMLElement;
+    return modal.current;
+  };
+
+  const handleMounted = () => {
+    manager.mount(getModal(), { disableScrollLock });
+
+    // Fix a bug on Chrome where the scroll isn't initially 0.
+    if (modalRef.current) {
+      modalRef.current.scrollTop = 0;
+    }
+  };
+
+  const handleOpen = useEventCallback(() => {
+    const resolvedContainer = getContainer(container) || getDoc().body;
+
+    manager.add(getModal(), resolvedContainer);
+
+    // The element was already mounted.
+    if (modalRef.current) {
+      handleMounted();
+    }
+  });
+
+  const isTopModal = React.useCallback(() => manager.isTopModal(getModal()), [manager]);
+
+  const handlePortalRef = useEventCallback((node: HTMLElement) => {
+    mountNodeRef.current = node;
+
+    if (!node) {
+      return;
+    }
+
+    if (open && isTopModal()) {
+      handleMounted();
+    } else if (modalRef.current) {
+      ariaHidden(modalRef.current, ariaHiddenProp);
```

Comment:
Can't we return the `aria-hidden` in `getRootProps` instead of using DOM methods directly?

Reply:
This will cause another re-render. I wouldn't change the logic of it now. I am just extracting the logic in a hook. We can revisit and see if we can improve in a later iteration.